File: customer-app/services/order-service/telemetry/telemetry.go
// Package telemetry initialises OpenTelemetry for order-service.
// Traces/metrics are exported to Platform App's existing OTel Collector —
// Customer App does not run its own observability stack. The resource
// attributes below are how Platform App's AI agents tell Customer App
// signals apart from Platform App's own.
package telemetry

import (
	"context"
	"errors"
	"fmt"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

// Init sets up the global TracerProvider and MeterProvider and returns a
// shutdown function.
// endpoint: OTel Collector gRPC address (Platform App's collector)
// serviceName: value of OTEL_SERVICE_NAME, e.g. "order-service"
// env: "dev" or "qa"
func Init(ctx context.Context, serviceName, endpoint, env string) (func(context.Context) error, error) {
	conn, err := grpc.NewClient(endpoint,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		return nil, fmt.Errorf("dial otel collector %s: %w", endpoint, err)
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
			semconv.DeploymentEnvironment(env),
			// service.namespace is the tag Platform App's AI agents use to
			// segregate Customer App signals from its own — see
			// docs/superpowers/specs/2026-08-15-platform-customer-app-split-notes.md
			semconv.ServiceNamespace("customer-app"),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("create otel resource: %w", err)
	}

	// ── Traces ──────────────────────────────────────────────────────────────
	traceExporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, fmt.Errorf("create otlp trace exporter: %w", err)
	}
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExporter),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)
	otel.SetTracerProvider(tp)

	// ── Metrics ─────────────────────────────────────────────────────────────
	metricExporter, err := otlpmetricgrpc.New(ctx, otlpmetricgrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, fmt.Errorf("create otlp metric exporter: %w", err)
	}
	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExporter)),
		sdkmetric.WithResource(res),
	)
	otel.SetMeterProvider(mp)

	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	shutdown := func(ctx context.Context) error {
		var errs []error
		if err := tp.Shutdown(ctx); err != nil {
			errs = append(errs, fmt.Errorf("tracer provider shutdown: %w", err))
		}
		if err := mp.Shutdown(ctx); err != nil {
			errs = append(errs, fmt.Errorf("meter provider shutdown: %w", err))
		}
		if err := conn.Close(); err != nil {
			errs = append(errs, fmt.Errorf("grpc conn close: %w", err))
		}
		return errors.Join(errs...)
	}

	return shutdown, nil
}

File: customer-app/services/catalog-service/app/telemetry.py
import logging

from opentelemetry import metrics, trace
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.propagate import set_global_textmap
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import SERVICE_NAME, SERVICE_NAMESPACE, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

logger = logging.getLogger(__name__)


def setup_telemetry(service_name: str, otlp_endpoint: str, env: str) -> None:
    # service.namespace=customer-app is how Platform App's collector/agents
    # segregate Customer App signals from its own — see
    # docs/superpowers/specs/2026-08-15-platform-customer-app-split-notes.md
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_NAMESPACE: "customer-app",
        "deployment.environment": env,
    })

    try:
        tp = TracerProvider(resource=resource)
        tp.add_span_processor(
            BatchSpanProcessor(OTLPSpanExporter(endpoint=otlp_endpoint, insecure=True))
        )
        trace.set_tracer_provider(tp)
    except Exception as exc:
        logger.warning("OTel trace exporter init failed (non-fatal): %s", exc)
        trace.set_tracer_provider(TracerProvider(resource=resource))

    try:
        mp = MeterProvider(
            resource=resource,
            metric_readers=[
                PeriodicExportingMetricReader(
                    OTLPMetricExporter(endpoint=otlp_endpoint, insecure=True)
                )
            ],
        )
        metrics.set_meter_provider(mp)
    except Exception as exc:
        logger.warning("OTel metric exporter init failed (non-fatal): %s", exc)
        metrics.set_meter_provider(MeterProvider(resource=resource))

    set_global_textmap(TraceContextTextMapPropagator())
    logger.info("OpenTelemetry initialised: service=%s endpoint=%s", service_name, otlp_endpoint)


def get_tracer() -> trace.Tracer:
    return trace.get_tracer("catalog-service")


def get_meter() -> metrics.Meter:
    return metrics.get_meter("catalog-service")
